### PR TITLE
Move cursor in conpty correctly after a backspace when we've delayed an EOL wrap

### LINKED ISSUE #1245: linewrap then backspace produces unexpected visuals, missing characters
# Environment

Windows build number: Microsoft Windows [Version 10.0.18362.175]
Windows Terminal version (if applicable): built using d82eab44d076e441f0b630d9b53a19dfa884c3ab

# Steps to reproduce

* open terminal with following profile:
```json
{
    "globals" : 
    {
        "alwaysShowTabs" : false,
        "defaultProfile" : "{9d2406b9-7e40-4e2b-a20d-82dc61c4cc24}",
        "initialCols" : 80,
        "initialRows" : 60,
        "requestedTheme" : "dark",
        "showTabsInTitlebar" : true,
        "showTerminalTitleInTitlebar" : false
    },
    "profiles" : 
    [
        {
            "acrylicOpacity" : 0.80000001192092896,
            "closeOnExit" : true,
            "colorScheme" : "Solarized Dark",
            "commandline" : "cmd.exe",
            "cursorColor" : "#FFFFFF",
            "cursorShape" : "bar",
            "fontFace" : "Consolas",
            "fontSize" : 10,
            "guid" : "{9d2406b9-7e40-4e2b-a20d-82dc61c4cc24}",
            "historySize" : 9001,
            "name" : "cmd",
            "padding" : "0, 0, 0, 0",
            "snapOnInput" : true,
            "startingDirectory" : "%USERPROFILE%",
            "useAcrylic" : true
        }
    ]
}
```

* input characters until the line wraps, backspace once, and enter an additional character

# Expected behavior

* Cursor should return to the previous position following the last character on the previous line after hitting backspace.  
* Additional input character (sent after backspace) should show up after the last character on the previous line.

# Actual behavior

* Cursor returns to previous line, but with a whitespace between the last character input and the cursor position
* Additional input character does not show up in terminal, although it is recognized

'..789ab' is received as input in the example below, even though it is not shown:
GIF of issue

### PR DESCRIPTION
**Pull request by zadjii-msft**
## Summary of the Pull Request

This is a fix that technically was caused by #357, though we didn't have the Terminal at the time, so I only fixed conhost then. When a client app prints the very last column in the buffer, the cursor is often not _actually_ moved to the next row quite yet. The cursor usually just "floats" on the last character of the row, until something happens. This could be a printable character, which will print it on the next line, or a newline, which will move the cursor to the next line manually, or it could be a backspace, which might take the cursor back a character. 

Conhost and gnome-terminal behave slightly differently here, and wt behaves differently all together. Heck, conhost behaves differently depending on what output mode you're in. 

The scenario in question is typing a full row of text, then hitting backspace to erase the last char of the row.

What we were emitting before in this case was definitely wrong - we'd emit a space at that last row, but then not increment our internal tracker of where the cursor is, so the cursor in conpty and the terminal would be misaligned. The easy fix for this is to make sure to always update the `_lastText` member appropriately. This is the `RightExclusive` change.

The second part of this change is to not be so tricksy immediately following a "delayed eol wrap". When we have just printed the last char like that, always use the VT sequence CUP the next time the cursor moves. Depending on the terminal emulator and it's flags, performing a BS in this state might not bring the cursor to the correct position. 

## References

#405, #780, #357

## PR Checklist
* [x] Closes #1245
* [x] I work here
* [ ] Tests added/passed 
* [n/a] Requires documentation to be updated

## Detailed Description of the Pull Request / Additional comments

With the impending #405 PR I have, this still works, but the sequences that are emitted change, so I didn't write a test for this currently.

## Validation Steps Performed

Tried the scenario for both #357 and #1245 in inception, `gnome-temrinal` and `wt` all, and they all display the cursor correctly.

### CODE REVIEW

**Comment by DHowett-MSFT** [comment]
Does this fix the version of #3277 that came back ☹️

**Comment by zadjii-msft** [comment]
> 
> 
> Does this fix the version of #3277 that came back ☹️

Haven't the faintest clue. I'd _probably_ doubt it, do we have a consistent repro for that one?

File: src/renderer/vt/paint.cpp [hunk 1]
```diff
@@ -461,10 +461,24 @@ using namespace Microsoft::Console::Types;
...
+    //
+    // GH#1245: This needs to be RightExclusive, _not_ inclusive. Otherwise, we
+    // won't update our internal cursor position tracker correctly at the last
+    // character of the row.
+    if (_lastText.X < _lastViewport.RightExclusive())
     {
         _lastText.X += static_cast<short>(columnsActual);
     }
+    // GH#1245: If we wrote the exactly last char of the row, then we're
+    // probably in the "delayed EOL wrap" state. Different terminals (conhost,
```
Comment (miniksa) [inline src/renderer/vt/paint.cpp]: I don't like the "probably" here. How could we be sure or how does it vary?

File: src/renderer/vt/paint.cpp [hunk 2]
```diff
@@ -461,10 +461,24 @@ using namespace Microsoft::Console::Types;
...
+    // gnome-terminal, wt) all behave differently with how the cursor behaves at
+    // an end of line. Marke that we're in the delayed EOL wrap state - we don't
```
Comment (miniksa) [inline src/renderer/vt/paint.cpp]:
```suggestion
    // an end of line. Mark that we're in the delayed EOL wrap state - we don't

```

File: src/renderer/vt/paint.cpp (hunk 1 above)
Comment (zadjii-msft) [reply src/renderer/vt/paint.cpp]: Yea this shouldn't say "probably", we _are_ in that state.

**Pull request merged**